### DISCUSSION TITLE
Misspelled Method in art.defences.trainer.certified_adversarial_trainer_pytorch.py

### DISCUSSION THREAD

**Discussion by GreedyGroot** [Trusted-AI/adversarial-robustness-toolbox#2032] [Q&A]
In art.defences.trainer.certified_adversarial_trainer_pytorch.py in line 213 the method that should be called is named _apply_preprocessing not apply_preprocessing resulting in an error.

Edit: Also the method set_foward_mode results in an error if I run it on PyTorch since a PyTorch model has no method set_foward_mode

**Comment by GiulioZizzo** [comment]
Hi @GreedyGroot  thank you for your interest in ART!

It sound like you may be passing a pytorch model (or maybe an incorrect ART estimator) to make the trainer. 

Essentially, AdversarialTrainerCertifiedPytorch requires a PytorchDeepZ estimator. So your python script should do something like 1) create a pytorch model, 2) pass the model to PytorchDeepZ to get a estimator with certification functionality, and 3) finally pass this estimator to AdversarialTrainerCertifiedPytorch. 

Have you been able to run the notebook here: https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/notebooks/certified_adversarial_training.ipynb ?

Specifically in relation to the two items you raised: 

- Correct, in general a PyTorch model will not have set_foward_mode. We make this method in the PytorchDeepZ estimator. 

- To avoid accessing protected members (and needing to disable pylint warnings) PytorchDeepZ has a wrapper around _apply_preprocessing. 

You could have a look at the PytorchDeepZ file itself to see what is going on behind the scenes:   

https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/art/estimators/certification/deep_z/pytorch.py

Let me know if the above resolves your errors! If this was the problem we could incorporate a check into AdversarialTrainerCertifiedPytorch to assert that the estimator being passed in is of the correct type, and if not raising a more informative error.  

Giulio
